share/completions/p4.fish

```diff
@@ -0,0 +1,803 @@
+# Fish completions for Perforce (p4) SCM
+# Based on the list of commands from
+# https://www.perforce.com/perforce/r16.1/manuals/cmdref/
+# Author: Boris Aranovich https://github.com/nomaed
+
+#########################################################
+### p4 command parsing
+#########################################################
+
+function __fish_print_p4_client_name
+    set -l p4info (p4 info -s 2> /dev/null)
+    if test -z "$p4info"
```

Comment:
One of the benefits of `string` is that it's a builtin, so there's basically no startup time. That makes it great for operations on very few or very small strings (for bigger things, `grep` and `sed` et al are still faster).

A `string` on nothing should be _quite_ fast. In fact it seems to be less than a millisecond (execute `string match SOMETHING; echo $CMD_DURATION`).

Usually, you don't need to care about the time `string` or `test` need, because usually the startup time of external tools is far greater. E.g. here, I'd estimate the `p4` call to be by far the dominating call.